### PR TITLE
Add VGG-16, SVDNet and SVRNet Models/Training Interface

### PR DESCRIPTION
**Pull request by meera-m-t**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,31 @@
-# classification-COVID-19_Radiography_Dataset
\ No newline at end of file
+# COVID-19 detection method based on SVRNet and SVDNet in lung x-rays
+This repository presents the implemntaion of SVRNet and SVDNet models in the paper [COVID-19 detection method based on SVRNet and SVDNet in lung x-rays](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8404611/pdf/JMI-008-017504.pdf).
```
Comment (umesh-timalsina) [inline README.md]:
```suggestion
This repository presents the implementation of `SVRNet` and `SVDNet` models from the paper [COVID-19 detection method based on SVRNet and SVDNet in lung x-rays](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8404611/pdf/JMI-008-017504.pdf).
```

File: README.md [hunk 2]
```diff
@@ -1 +1,31 @@
...
+
+![alt text](images/model.png)
```
Comment (umesh-timalsina) [inline README.md]:
```suggestion
![model](images/model.png)
```

File: README.md [hunk 3]
```diff
@@ -1 +1,31 @@
...
+
+
+## Data
+COVID-19 Radiography Database (COVID-19 Chest X-ray Database)
```
Comment (umesh-timalsina) [inline README.md]:
```suggestion
## Data: COVID-19 Radiography Database (COVID-19 Chest X-ray Database)
```

File: README.md [hunk 4]
```diff
@@ -1 +1,31 @@
...
+
+You can download this dataset from [here](https://www.kaggle.com/tawsifurrahman/covid19-radiography-database?fbclid=IwAR3JBdbiHVJFYHcNlR3r3Z1esKY3UKrCHJd8Nrhv4OPXdGhOZWtEcqtjEEg). This article selects 1560 lung x-rays from them as training set, validation set, and test set, including 1341 lung x-ray images that tested negative for COVID-19 and 219 lung x-ray images that tested positive for COVID-19. Before conducting the classification experiment on whether images contain COVID-19, the data set is divided into two categories. One type is used to train the model, including training set and validation set, which respectively account for 60% and 20% of the total data set; the other type is used to test the performance of the model, which accounts for 20% of the total data set. Image Preprocessing: OpenCV mainly relies on the cv2.imread() function to read the image, which can read the x-ray image from the original file. However, the contrast of the image at this time is unsatisfactory, which will affect the detection accuracy, so this article uses the limited contrast adaptive histogram equalization (CLAHE). Run:
```
Comment (umesh-timalsina) [inline README.md]: What do you mean by `This article`?

Too Much Details

File: README.md [hunk 5]
```diff
@@ -1 +1,31 @@
...
+
+```bash
+$ python preprocessing_data.py
```
Comment (umesh-timalsina) [inline README.md]: This should come after you install the package requirements imo.

File: README.md [hunk 6]
```diff
@@ -1 +1,31 @@
...
+
+```bash
+$ pip install -r requirements.txt
+```
+
+## Model Training/Testing 
+To Train the model, choose which model you want to train from main.py 
+
+```bash
+python --model_name ['VGG16', 'SVRNet', 'SVDNet']  main.py
```
Comment (umesh-timalsina) [inline README.md:17]: Print the contents of `python main.py --help` here.

File: README.md [hunk 7]
```diff
@@ -1 +1,31 @@
...
+```
+
+
+
```
Comment (umesh-timalsina) [inline README.md]: Too Many empty lines. Just leave one, its a standard practice.

File: main.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                    batch_size=32,                                   
+                    image_size=(IMG_SIZE, IMG_SIZE),
+                    validation_split=None,
+                    subset=None,
+                    seed=1
+                )
+    return test_data.cache().prefetch(buffer_size=AUTOTUNE)         
+
+def main():
+    parser = argparse.ArgumentParser(description='add, modify and delete upstream nodes')
```
Comment (umesh-timalsina) [inline main.py:65]: Can we change the description to something like?
```py
    parser = argparse.ArgumentParser(description='Train different models using the COVID-19 datasets')
```

File: main.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    callback = EarlyStopping(monitor='categorical_accuracy', patience=10)
+    history = model.fit(
+        train_data,    
+        validation_data=val_data,  
+        epochs=EPOCHS,
+        callbacks=callback,
+        batch_size=BATCH_SIZE,  
+        shuffle=True
+        )
+    print(history.history.keys())
```
Comment (umesh-timalsina) [inline main.py:149]: Why is this print statement here?
```py
    print(history.history.keys())
```

File: main.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+    # summarize history for accuracy
+    plot1 = plt.figure(1)
+    plt.plot(history.history['categorical_accuracy'])
+    plt.plot(history.history['val_categorical_accuracy'])
+    plt.title('model accuracy')
+    plt.ylabel('accuracy')
+    plt.xlabel('epoch')
+    plt.legend(['train', 'val'], loc='upper left')
+    plt.savefig('SVDNet_model_accuracy.png')
```
Comment (umesh-timalsina) [inline main.py:176]: Change this to select the appropriate model name

File: main.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+    # # summarize history for loss
+    plot2 = plt.figure(2)
+    plt.plot(history.history['loss'])
+    plt.plot(history.history['val_loss'])
+    plt.title('model loss')
+    plt.ylabel('loss')
+    plt.xlabel('epoch')
+    plt.legend(['train', 'val'], loc='upper left')
+    plt.savefig(f'{model_name}_model_lost.png')
```
Comment (umesh-timalsina) [inline main.py:186]:
```suggestion
    plt.savefig(f'{model_name}_model_loss.png')
```

File: main.py [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+    ax.yaxis.set_ticklabels(['False','True'])
+    fig = ax.get_figure()
+    fig.savefig(f'{model_name}_model_confusion_matrix.png')
+
+
+if __name__ == "__main__":
+    main()
+
+
+
```
Comment (umesh-timalsina) [inline main.py:208]: Too many blank lines, just use one.

File: preprocessing_data.py [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import cv2
+import glob
+import os
+import matplotlib.pyplot as plt
+import numpy as np
+import splitfolders
+for name in ['COVID',  'Normal']:
+    for file in glob.glob(f"COVID-19_Radiography_Dataset/{name}/*.png"):
```
Comment (umesh-timalsina) [inline preprocessing_data.py]: Just a suggestion. Change this to be generic as well?

File: preprocessing_data.py [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+        name_file = os.path.basename(file)        
+        image = cv2.imread(file)
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
+        equalized = clahe.apply(gray)    
+        cv2.imwrite(f"COVID-19_Radiography_Dataset/preprocessed_data/{name}/{name_file}", equalized)
```
Comment (umesh-timalsina) [inline preprocessing_data.py]: Same comment as above.

File: preprocessing_data.py [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+
+
+splitfolders.ratio('COVID-19_Radiography_Dataset/preprocessed_data', output="output", seed=1337, ratio=(.8, 0.2))
```
Comment (umesh-timalsina) [inline preprocessing_data.py]: Same comment as above

**Review by umesh-timalsina — CHANGES_REQUESTED** [review]
Good Work Overall. I have left some comments through a quick 15 mins review. 

It might be nice to publish model weights in google drive? What do you think?

File: README.md (hunk 1 above)
Comment (sameerahtalafha) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (sameerahtalafha) [reply README.md]: done

File: README.md (hunk 4 above)
Comment (sameerahtalafha) [reply README.md]: done

File: README.md (hunk 5 above)
Comment (sameerahtalafha) [reply README.md]: done

File: main.py (hunk 11 above)
Comment (sameerahtalafha) [reply main.py:186]: done

File: main.py (hunk 12 above)
Comment (sameerahtalafha) [reply main.py:208]: done

File: preprocessing_data.py (hunk 13 above)
Comment (sameerahtalafha) [reply preprocessing_data.py]: done

File: preprocessing_data.py (hunk 14 above)
Comment (sameerahtalafha) [reply preprocessing_data.py]: done

File: preprocessing_data.py (hunk 15 above)
Comment (sameerahtalafha) [reply preprocessing_data.py]: done
